server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -449,6 +449,29 @@ public Tuple<ShardRouting, ShardRouting> relocateShard(
         return Tuple.tuple(source, target);
     }
 
+    public void relocateOrReinitializeShard(
+        ShardRouting startedShard,
+        String nodeId,
+        long expectedShardSize,
+        RoutingChangesObserver changes
+    ) {
+        if (startedShard.isSearchable() == false) {
+            remove(startedShard);
+            var unassignedInfo = new UnassignedInfo(UnassignedInfo.Reason.REINITIALIZED, "relocating unsearchable shard");
+
+            var assignedShards = assignedShards(startedShard.shardId());
+            var promotableShard = assignedShards.stream().filter(ShardRouting::isPromotableToPrimary).findAny();
+            assert promotableShard.isEmpty() : "multiple promotable shards are not supported yet";
+            for (ShardRouting replica : List.copyOf(assignedShards)) {
```

Comment:
It is probably needed, but not obvious to me why, perhaps you can explain why we need to also reinit the search copies?